### DISCUSSION TITLE
The rerank model cannot run on the GPU, causing it to be very slow.

### DISCUSSION THREAD

**Discussion by zero456** [open-webui/open-webui#6530] [General]
Computer:
  2X Intel(R) Xeon(R) Gold 6242R CPU
  64.0 GB RAM
  NVIDIA Quadro RTX 6000

Open-WebUI settings:
  Engine: Ollama
  Embedding Batch Size = 12
  Hybrid Search: Enabled
  Embed Model: bge-m3:latest
  Rerank Model: baai/bge-reranker-v2-m3 (downloaded from Hugging Face)

After running the query, from the backend, we observed that the embedding process completes very quickly, with brief CUDA GPU utilization. Then, the CPU utilization increases significantly to 60%~100% and remains high for a prolonged period until the answer is generated.
Based on these observations, we suspect that the rerank model is running on the CPU. Is it possible to modify it to run on the GPU to improve the speed?

**Comment by ciprian-negrila** [comment]
This problem comes from the way the device type configuration is assigned to the sentence transformers. 

This needs to be fixed in a sustainable way in the future. My recommendation to the team is to implement a similar drop-down for rerank as we have for embedding and use Ollama. To fix this problem in the interim you can hardcode using CUDA instead of CPU in the sentence_transformers settings when configuring the model. DO NOT DO THIS unless you have a cuda GPU and YOU KNOW it works for inference on Ollama for example.

You will need to edit one line of code in your openwebui package. How to get sentence_transformers to work correctly on GPU:
In your environment go to: **Lib\site-packages\open_webui\apps\retrieval**
Edit the **main.py** file and change this code:

```
        else:
            import sentence_transformers

            try:
                app.state.sentence_transformer_rf = sentence_transformers.CrossEncoder(
                    get_model_path(reranking_model, auto_update),
                    device=DEVICE_TYPE,
                    trust_remote_code=RAG_RERANKING_MODEL_TRUST_REMOTE_CODE,
                )
            except:
                log.error("CrossEncoder error")
                app.state.sentence_transformer_rf = None
                app.state.config.ENABLE_RAG_HYBRID_SEARCH = False
```

To this:

```
        else:
            import sentence_transformers

            try:
                app.state.sentence_transformer_rf = sentence_transformers.CrossEncoder(
                    get_model_path(reranking_model, auto_update),
                    device='cuda',
                    trust_remote_code=RAG_RERANKING_MODEL_TRUST_REMOTE_CODE,
                )
            except:
                log.error("CrossEncoder error")
                app.state.sentence_transformer_rf = None
                app.state.config.ENABLE_RAG_HYBRID_SEARCH = False
```

Basically you need to replace in the rerank model config update function the DEVICE_TYPE variable with hardcoded 'cuda' text instead.

Keep in mind the following:
- If and when you update open-webui this file will be rewritten and you will have to make the change again (until properly fixed)
- You need to change the file, start open-webui, update your rerank model config (remove it and add it again for example or replace it and then change it back), stop open-webui and then start again to use the newly saved config using GPU.
- Profit
- ???

**Reply by Koesn** [reply]
It works very fast. Thank's for this.

**Comment by ciprian-negrila** [comment]
By the way, I have this problem because somehow even after installing torch correctly (AFAIK) and I have a CUDA device, my torch still returns none:

```
nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2024 NVIDIA Corporation
Built on Fri_Jun_14_16:44:19_Pacific_Daylight_Time_2024
Cuda compilation tools, release 12.6, V12.6.20
Build cuda_12.6.r12.6/compiler.34431801_0
```

But in python:
```
python
Python 3.11.10 | packaged by conda-forge | (main, Sep 30 2024, 17:50:51) [MSC v.1941 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> print(torch.cuda.is_available())
False
>>>
```

So most likely this is a problem with torch installs and packages being mixed up OR some local config problem. Still, Ollama and transformers work fine on CUDA, if you force it as explained above.

**Comment by mahenning** [comment]
@zero456 Do you use docker? If yes, there is a [cuda image](https://github.com/orgs/open-webui/packages/container/open-webui/318633120?tag=main-cuda) of open-webui. AFAIK, if you use the default image, cuda is not used within the open-webui container.

If you don't use docker, and install open-webui with pip, it should install the appropiate cuda drivers into the current virtual environment. But I tested it, and to let the reranker (or and other whisper/retrieve model I suppose) use the GPU, you have to set the environment variable `USE_CUDA_DOCKER=true`. You can set it with `export USE_CUDA_DOCKER=true` in the current terminal, or e.g. write it in your `.bashrc`. I think you can set it even if you already configured open-webui. But you have to restart the open-webui service to use the GPU after setting the env var. 
I'm not sure where exactly it's used, but at least [at this line](https://github.com/open-webui/open-webui/blob/29a271959556743e6deb4d55a5a982983335d7ab/backend/open_webui/env.py#L37) it checks for the `USE_CUDA_DOCKER` variable to set the `DEVICE_TYPE` variable, which is later used, and which @ciprian-negrila overwrites in their fix.

I think this is rarher unintuitive to set a variable with "DOCKER" in its name to use GPU outside of docker. But it seems that the focus lays on usability with docker, and the pip installation is more of an afterthought. The pip installation is also one version behind (current: 0.4.7) the latest version (0.4.8). Maybe there is a better way to use GPU, but I didn't find it.

----
Edit: I missed your mention that you use ollama for your embedding model. I assume your ollama instance has GPU access. Unfortunatelly, it seems that the reranking model is fixed as a sentence-transformer model, which is managed by open-webui itself. And open-webui by default has no GPU access, which can be fixed by the steps I mentioned above.

**Reply by richtong** [reply]
That's really interesting on a Mac at least with OpenWebUI runing a python pip package I have exactly the same problem. I can get ollama to run the embedding but the rerankeer is always run on the CPU

There is no way that I can see to force this. This seems to be true with the sentence transformers in general. I'll post a separate issue on this

**Reply by fdkgenie** [reply]
> That's really interesting on a Mac at least with OpenWebUI runing a python pip package I have exactly the same problem. I can get ollama to run the embedding but the rerankeer is always run on the CPU
> 
> There is no way that I can see to force this. This seems to be true with the sentence transformers in general. I'll post a separate issue on this

maybe we can try to hardcode device='mps' for Apple M silicon chip?

**Reply by fdkgenie** [reply]
> > That's really interesting on a Mac at least with OpenWebUI runing a python pip package I have exactly the same problem. I can get ollama to run the embedding but the rerankeer is always run on the CPU
> > There is no way that I can see to force this. This seems to be true with the sentence transformers in general. I'll post a separate issue on this
> 
> maybe we can try to hardcode device='mps' for Apple M silicon chip?

I tried and got error, what a dirty hack!

**Reply by apunkt** [reply]
pip install open-webui 0.6.8 using `USE_CUDA_DOCKER=true` finally put the reranker on GPU.
Thanks for this! I was looking for a long time to get it running, the DOCKER part in the name is confusing...

**Comment by richtong** [comment]
I was thinking about trying the exact same thing. So many times just setting device=cuda just works. So a good try to just do that.

Any other ideas and where did you set this. It's in the RAG code right.

I guess I could also try a pipeline but would be nice if it just works Out of the box. :-)

Sent via Superhuman iOS ( ***@***.*** )

**Comment by richtong** [comment]
Who works on the Mac version. I can submit it an issue and look at the rag code.

Sent via Superhuman iOS ( ***@***.*** )

**Reply by fdkgenie** [reply]
You're running on docker or not?

**Comment by richtong** [comment]
No docker. Venv with open-webui pipx installed.

Sent via Superhuman iOS ( ***@***.*** )

**Reply by fdkgenie** [reply]
I got it, just change "cpu" to "mps" and it works quite fast on native M3pro.
The last time I tried without success, it was because I ran on a docker, it's Linux based so it doesn't support MPS naturally.

**Comment by richtong** [comment]
That's great. Which file. I'm sure I can do some detection code.

Sent via Superhuman iOS ( ***@***.*** )

**Comment by richtong** [comment]
Sorry the code base is something i haven't had time to look at.

Sent via Superhuman iOS ( ***@***.*** )

**Comment by richtong** [comment]
OK I fixed this in `./backend/open-webui/env.py` and also note that the installation development instructions are out of date, you should use uv instead of a pip install. I updated the README.md in backend to cover this although I can send a PR to update the documentation if you like

Here is the fix, this was the old code there (which I have to say I may have broken). Is it possible that open-webui is not built with PyTorch is that why there is this assertion and exception? Is that why there is exception throw. It is better to test for torch not loaded if it isn't, but the requirements.txt always asks for it:

```old
if USE_CUDA.lower() == "true":
     try:
        import torch
        assert torch.cuda.is_available(), "CUDA not available"
        DEVICE_TYPE = "cuda"
     except Exception as e:
        cuda_error = (
             "Error when testing CUDA but USE_CUDA_DOCKER is true. "
             f"Resetting USE_CUDA_DOCKER to false: {e}"
         )
         os.environ["USE_CUDA_DOCKER"] = "false"
         USE_CUDA = "false"
         DEVICE_TYPE = "cpu"
 else:
     DEVICE_TYPE = "cpu"
```


I did break up the if statement on docker since its a little complicated and also apparent some
installs do not include torch, so here you go
```new
try:
  import torch
exception as e:
  # no torch use coda
  DEVICE_TYPE = "cpu"
  USE_CUDA = "false"
else:
  # only use cuda in docker if it is allowed
  if torch.backends.mps.is_available() and torch.backends.mps.is_built():
     DEVICE_TYPE = "mps"
  elif torch.cuda.is_available() and torch.backends.cuda.is_built():
     if not DOCKER or (DOCKER and USE_CUDA.lower() == "true"):
         DEVICE_TYPE = "cuda"
  else:
    DEVICE_TYPE = "cpu"
    USE_CUDA = "false"
```

**Reply by richtong** [reply]
Net, net, I'm new to this project, is there a process for contributing to it. Should I create an issue and a PR?

**Reply by richtong** [reply]
And also FYI, this speeds up RAG by about 10x, it's like a blink of an eye now.

**Reply by richtong** [reply]
I stuffed the fixes at https://github.com/richtong/open-webui/tree/rich-mps and it ready to be PRed depending on how folks handle this here.

**Comment by richtong** [comment]
FYI this is a duplicate of https://github.com/open-webui/open-webui/discussions/8278 and I've issued a PR to fix

**Reply by richtong** [reply]
this seems to work on main and also dev branch, but I'm getting other RAG errors that I don't think are related. See other discussion on padding token

**Comment by Siegurd01** [comment]
Take look on similar solution https://github.com/open-webui/open-webui/discussions/8932

**Comment by Iyozhikov** [comment]
My systemd unit to enable RAG on GPU (I have several small P1000 with 4Gb vRAM):
```bash
cat /etc/systemd/system/open-webui.service
[Unit]
Description=Open WebUI Service
After=network-online.target

[Service]
WorkingDirectory=/opt/open-webui
ExecStart=/opt/venvs/open-webui/bin/open-webui serve --port 80
User=aiservice
Group=aiservice
Restart=always
RestartSec=3
Environment="PATH=$PATH"
Environment="USE_CUDA_DOCKER=true"
Environment="PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True,max_split_size_mb:512"
Environment="CUDA_VISIBLE_DEVICES=3"
AmbientCapabilities=CAP_NET_BIND_SERVICE

[Install]
WantedBy=multi-user.target
```
<img width="1876" height="1310" alt="image" src="https://github.com/user-attachments/assets/7fc884bd-83d3-48f2-9ce5-bd9549840421" />

**Reply by Iyozhikov** [reply]
Running this model: `nomic-ai/nomic-embed-text-v1.5         model             2.1G`
